add-bookings

Rat-In-Hat opened the pull request.

Rat-In-Hat committed:
feat

добавить репозиторий для юзера
добавить первую версию схему
изменить app.yml для корректной работы программы
добавить внутреннюю БД

Rat-In-Hat committed:
feat

создать репозиторий под вещи
реализовать функционал

Rat-In-Hat committed:
feat

реализовать часть функционала по бронированию вещей.
создать необходимые классы и модели.

Rat-In-Hat committed:
del

del db

Rat-In-Hat committed:
fix

gitIGNORE

Rat-In-Hat committed:
feat

реализовать весь функционал бронирования

Rat-In-Hat committed:
feat

реализовать весь функционал комментирования

Rat-In-Hat committed:
del

удалить реализацию imMemmory

Rat-In-Hat committed:
style

reformat code

Rat-In-Hat committed:
style

reformat code

Rat-In-Hat committed:
style

reformat code

Rat-In-Hat committed:
style

reformat code

Rat-In-Hat committed:
ref

add Transactional

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,12 +1,78 @@
+    @GetMapping
+    public ResponseEntity<List<BookingDtoOut>> getAllForCurrentUser(
+            @RequestHeader(requestHeader) Long userId,
+            @RequestParam(name = "state", defaultValue = "ALL") String state) {
```
Принимай параметр сразу как тип `State`.

Rat-In-Hat replied:
Решил иначе, ниже подробности

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,12 +1,78 @@
+            return ResponseEntity.ok(
+                    bookingService.getAllBookingsForCurrentUserId(userId, State.valueOf(state)));
+        } catch (IllegalArgumentException e) {
+            throw new BadRequestException(String.format("Unknown state: %s", state));
```
Не понимаю назначения этого `try/catch`.
Такое исключение выkетает из-за `State.valueOf(state)` ? 

Такая надобность или отпадет или сделай это в отдельном методе

Rat-In-Hat replied:
Сделал отедбныей метод. Тесты требуют особенной обработки этой ошибки. Сделал статик метод для преобразования из строки в State, он выкинет то, что просит постман тест. Что то еще городить не вижу смысла

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -1,20 +1,42 @@
-    private final Long id;
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
```
При обилии аннотаций всё сливается - делай отступы между полями.

Rat-In-Hat replied:
Это можно поставить как условие в код стайле? Не нашел такой опции, пока оставлю так.

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -1,20 +1,42 @@
+    /*Тыкался я мыкался, хотел сделать, чтобы отсюда сразу шла вставка по дефолту
+     * убил кучу времени но ничего не получилось и пришлось уже в сервисе передавать значение
+     * может подскажете способ?*/
+    @Enumerated(EnumType.STRING)
```
И что простое `private Status status = Status.WAITING;` не работает?
Ну выкинь ломбок, напиши все ручками - управляй классом самостоятельно.

Rat-In-Hat replied:
Ок

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,76 @@
+
+public interface BookingRepository extends JpaRepository<Booking, Long> {
+
+    /*Покажет, брал ли пользователь эту вещь в аренду и завершилась ли она*/
```
Оформляй комменты как `javaDoc`
Дело в том что он будет доступен при просмотре метода, это комбинация клавиш `ctrl+q`
Ну и в случае автогенерации документации и прочее...
Вот так это выглядит.

![image](https://github.com/ImING0/java-shareit/assets/3994663/e64fa605-e36b-45d9-ad9b-9e44b09aeb43)

Rat-In-Hat replied:
Классное решение, добавил в репозитории)

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/service/IBookingService.java:
```diff
@@ -0,0 +1,25 @@
+                         Boolean approved,
+                         Long userId);
+
+    BookingDtoOut getByBookingIdAndUserId(Long bookingId,
```
Мне не нравятся названия трех последних методов.
Не вижу сути...
Ну названия собраны как у репозитория, но тут же не он, тут можно написать понятно.

А между `getAllBookingsForCurrentUserId` и `getAllItemBookingsForOwnerId` вообще разницы не улавливаю... принимают одинаковые параметры и возвращают одинаковый список!

Rat-In-Hat replied:
Последние два метода, getAllBookingsForCurrentUserId и getAllItemBookingsForOwnerId, действительно принимают одинаковые параметры и возвращают похожие списки. Это связано с тем, что оба метода служат для получения списка бронирований, но они предназначены для разных видов пользователей: getAllBookingsForCurrentUserId - для текущего пользователя, а getAllItemBookingsForOwnerId - для владельца. Названия методов отражают эту разницу и обеспечивают ясность в том, какой тип бронирований ожидается при вызове каждого из них.

Однако, я готов улучшить читаемость кода, добавив поясняющие JavaDocs к этим методам. Это поможет разработчикам легче понимать, какие данные они могут ожидать в ответе. Если у вас есть предложения по изменению названий методов, я готов их рассмотреть.

Rat-In-Hat replied:
getByBookingIdAndUserId изменил на getBookingById

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/service/impl/BookingService.java:
```diff
@@ -0,0 +1,255 @@
+    private final UserRepository userRepository;
+    private final ItemRepository itemRepository;
+    private final BookingMapper bookingMapper;
+    private final UserMapper userMapper;
```
мапперы не используются...

Rat-In-Hat replied:
удалил, видимо использовались да перестали во время разработки)

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/service/impl/BookingService.java:
```diff
@@ -0,0 +1,255 @@
+        Item item = getItemOrThrowIfNotExist(bookingDtoIn.getItemId());
+        validateBookingBeforeCreate(userId, bookingDtoIn);
+        Booking booking = bookingMapper.toBooking(user, item, bookingDtoIn);
+        booking.setStatus(Status.WAITING);
```
В мапере вижу установку статуса в WAITING тут лишнее получается...

Rat-In-Hat replied:
Согласен, логичнее установить в слое сервисе у DTO IN статус, а потом уже маппить. так и сделал

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/mapper/ItemMapper.java:
```diff
@@ -1,18 +1,65 @@
-    public ItemDto toItemDto(Item item) {
+    public ItemDto toItemDtoWithoutBooking(Item item) {
+        return ItemDto.builder()
+                .id(item.getId())
```
Ну ваще не красиво!! :face_with_spiral_eyes: 

Код, повторяется 4-е раза, ну можно же с этим что-то сделать!

```
                .id(item.getId())
                .owner(item.getOwner())
                .name(item.getName())
                .description(item.getDescription())
                .available(item.getAvailable())
```

Rat-In-Hat replied:
Согласен, убожество еще то! Аж самому блевать захотелось, когда перечитал)

Rat-In-Hat replied:
поменять в общем логику. мапперы в топку, в сервисе теперь всё как надо

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+    List<Item> findAllByOwner(Long ownerId);
+
+    @Query("select i " + "from Item as i " + "where i.available = true and "
```
Оформляй запросы, набирая команды SQL в верхнем регистре: SELECT

Rat-In-Hat replied:
Сделал

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -1,15 +1,29 @@
+@NoArgsConstructor
+@AllArgsConstructor
 public class User {
+    // Мне кажется, что проверки на нулль тут избыточны, ведь в БД уже есть проверка на это?
```
Это как бы не проверка, а контракт.
Ты подсказываешь как генерить эти колонки в БД, в случае если оно происходит автоматом.
Ты подсказываешь разработчикам как оно должно быть создано в БД.

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

Rat-In-Hat committed:
docs

добавить немного документации в репозитории

Rat-In-Hat committed:
docs, ref

добавить пояснительную документацию к методам BookingService

изменить название метода

Rat-In-Hat committed:
ref

убрать установление статуса из маппера в слой сервиса

Rat-In-Hat committed:
ref

изменить регистр select

Rat-In-Hat committed:
ref

убрать ненужные мапперы, изменить логику получения предметов с бронированиями и без.

Rat-In-Hat committed:
ref

добавить метод fromString в State, теперь он выкинет ошибку, если передан не тот аргумент )
